starport-blog/01-index.md

```diff
@@ -0,0 +1,350 @@
+  "github.com/google/uuid"
+)
+var _ sdk.Msg = &MsgCreatePost{}
+type MsgCreatePost struct {
+  Creator sdk.AccAddress `json:"creator" yaml:"creator"`
+  Title   string         `json:"title" yaml:"title"`
+  ID      string         `json:"id" yaml:"id"`
+}
+```
+
+Similarly to the post struct, `MsgCreatePost` contains creator and title properties. We don’t include ID property, because `MsgCreatePost` defines only the data we accept from the user—we will be generating ID automatically on the next step.
+
+```go
+// NewMsgCreatePost creates the `MsgCreatePost` message
+func NewMsgCreatePost(creator sdk.AccAddress, title string) MsgCreatePost {
+  return MsgCreatePost{
+    Creator: creator,
+    Title:   title,
+    ID:      uuid.New().String(),
+  }
+}
+```
+
+`NewMsgCreatePost` is a constructor function that creates `MsgCreatePost` message. The following five functions have to be defined to implement `Msg` interface. They allow you to perform validation that doesn’t require access to store (like checking for empty values), etc.
+
+```go
+// Route ...
+func (msg MsgCreatePost) Route() string {
+  return RouterKey
+}
+// Type ...
+func (msg MsgCreatePost) Type() string {
+  return "CreatePost"
+}
+// GetSigners ...
+func (msg MsgCreatePost) GetSigners() []sdk.AccAddress {
+  return []sdk.AccAddress{sdk.AccAddress(msg.Creator)}
+}
+// GetSignBytes ...
+func (msg MsgCreatePost) GetSignBytes() []byte {
+  bz := ModuleCdc.MustMarshalJSON(msg)
+  return sdk.MustSortJSON(bz)
+}
+// ValidateBasic ...
+func (msg MsgCreatePost) ValidateBasic() error {
+  if msg.Creator.Empty() {
+    return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "creator can't be empty")
+  }
+  return nil
+}
+```
+
+Going back to `GetCmdCreatePost`, you created a new `MsgCreatePost` message and broadcasted it with `GenerateOrBroadcastMsgs`.
+
+After being broadcasted, the messages are processed by handlers.
+
+## x/blog/handler.go
+
+```go
+import (
+  // Existing imports...
+  "github.com/example/blog/x/blog/types"
+)
+```
+
+You should already have `func NewHandler` defined which lists all available handlers. Modify it to include a new function called `handleMsgCreatePost`.
+
+```go
+    switch msg := msg.(type) {
+    case MsgCreatePost:
+      return handleMsgCreatePost(ctx, k, msg)
+    default:
+```
+
+Now let’s define `handleMsgCreatePost`:
+
+```go
+func handleMsgCreatePost(ctx sdk.Context, k Keeper, msg MsgCreatePost) (*sdk.Result, error) {
+  var post = types.Post{
+    Creator: msg.Creator,
+    ID:      msg.ID,
+    Title:   msg.Title,
+  }
+  k.CreatePost(ctx, post)
+  return &sdk.Result{Events: ctx.EventManager().Events()}, nil
+}
+```
+
+In this handler you create a `Post` object (post type was defined in the very first step). You populate the post object with creator and title from the message (`msg.Creator` and `msg.Title`) and generate a unique ID with `uuid.New().String()`. Make sure to add `"github.com/google/uuid"` to the import at the top of this file.
```

Comment:
is `uuid.New()` deterministic?

Reply:
```suggestion
In this handler you create a `Post` object (post type was defined in the very first step). You populate the post object with creator and title from the message (`msg.Creator` and `msg.Title`) and use the unique ID that was generated in `tx.go` with `NewMsgCreatePost()` using `uuid.New().String()`.
```

Reply:
> is uuid.New() deterministic?

It is not, but it is used during message creation, so I didn't see a problem in that.

Reply:
if it's created in message generation you're right there's no problem, but is there also a check happening to confirm uniqueness?